README.md
# koroFileHeader 

<a href="https://github.com/OBKoro1/koro1FileHeader">
    <img alt="koro1FileHeader Repo stars" src="https://img.shields.io/github/stars/OBKoro1/koro1FileHeader">
</a>
<a href="https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE%E5%AD%97%E6%AE%B5">
    <img alt="wiki文档详细" src="https://img.shields.io/badge/wiki文档-齐全详细-blue">
</a>
<a href="https://github.com/OBKoro1/koro1FileHeader/wiki/%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97">
    <img alt="持续维护" src="https://img.shields.io/badge/2018年开源-持续维护-blue">
</a>
<a href="https://github.com/OBKoro1/koro1FileHeader/releases">
    <img alt="cicd" src="https://img.shields.io/badge/版本打包-release-blue">
</a>
<a href="https://github.com/OBKoro1/koro1FileHeader/blob/master/LICENSE">
    <img alt="开源协议-MIT" src="https://img.shields.io/badge/license-MIT-blue">
</a>

### 简介

1. VSCode插件: 用于一键生成文件头部注释并自动更新最后编辑人和编辑时间、函数注释自动生成和参数提取。
2. 插件可以帮助用户养成良好的编码习惯，规范整个团队风格。
3. 从2018年5月维护至今, 关闭issue 500+ ，拥有39.7w+的用户，VSCode图表统计日均安装200-500
4. 经过多版迭代后，插件支持所有主流语言,灵活方便，文档齐全，食用简单！
5. 觉得插件不错的话，就给个[Star](https://github.com/OBKoro1/koro1FileHeader)⭐️吧~

### 主要功能

1. 自动生成文件头部注释，自动更新最后编辑人、最后编辑时间等。
2. 一键生成函数注释，支持函数参数自动提取并列到注释中。
3. 支持添加佛祖保佑永无bug、神兽护体、甩葱少女等好玩有趣的图像注释
4. 配置非常灵活方便，各种细节都能配置，可以量身打造适合你的注释。
4. 支持所有主流语言, 配置文档非常详细，齐全。

### 使用效果：

**头部注释和注释图案**

![example.gif](https://raw.githubusercontent.com/OBKoro1/koro1FileHeader/master/images/example.gif)

**函数注释: 自动提取函数参数**

1. 支持主流语言自动提取函数参数

![koroFileHeader函数参数提取](https://github.com/OBKoro1/koro1FileHeader/blob/master/images/function-params.gif?raw=true)

2. 多行函数参数支持自动提取

![多行函数参数自动提取参数](https://github.com/OBKoro1/koro1FileHeader/blob/dev/images/docs/multiLineParamsCreate.gif?raw=true)

**注释图案**

[支持一键添加佛祖保佑永无BUG、神兽护体等注释图案](https://github.com/OBKoro1/koro1FileHeader/wiki/%E4%BD%9B%E7%A5%96%E4%BF%9D%E4%BD%91%E6%B0%B8%E6%97%A0BUG%E3%80%81%E7%A5%9E%E5%85%BD%E6%8A%A4%E4%BD%93%E3%80%81%E6%B3%A8%E9%87%8A%E5%9B%BE%E6%A1%88)

![添加注释图案](https://github.com/OBKoro1/koro1FileHeader/raw/master/images/codeDesign.gif?raw=true)

### 安装和快速上手

[安装和快速上手](https://github.com/OBKoro1/koro1FileHeader/wiki/%E5%AE%89%E8%A3%85%E5%92%8C%E5%BF%AB%E9%80%9F%E4%B8%8A%E6%89%8B)

### 快速查看插件功能与配置方法

**功能目录**：查看[配置](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE)右侧自动生成的markdown目录，浏览所有最新最全的功能。

**配置简介**：[配置字段](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE%E5%AD%97%E6%AE%B5)

## Wiki文档

[安装和快速上手](https://github.com/OBKoro1/koro1FileHeader/wiki/%E5%AE%89%E8%A3%85%E5%92%8C%E5%BF%AB%E9%80%9F%E4%B8%8A%E6%89%8B)

[支持语言](https://github.com/OBKoro1/koro1FileHeader/wiki/%E6%94%AF%E6%8C%81%E8%AF%AD%E8%A8%80)

[插件设置/配置](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE)

[配置字段](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE%E5%AD%97%E6%AE%B5)

[更新日志](https://github.com/OBKoro1/koro1FileHeader/wiki/%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97)

[常见问题](https://github.com/OBKoro1/koro1FileHeader/wiki/%E5%B8%B8%E8%A7%81%E9%97%AE%E9%A2%98)

[佛祖保佑永无BUG、神兽护体等注释图案](https://github.com/OBKoro1/koro1FileHeader/wiki/%E4%BD%9B%E7%A5%96%E4%BF%9D%E4%BD%91%E6%B0%B8%E6%97%A0BUG%E3%80%81%E7%A5%9E%E5%85%BD%E6%8A%A4%E4%BD%93%E3%80%81%E6%B3%A8%E9%87%8A%E5%9B%BE%E6%A1%88)


### 我的其他开源推荐
#### [stop-mess-around](https://github.com/OBKoro1/stop-mess-around)

减少摸鱼的时间和频率的Chrome插件：**在上班/学习期间很容易下意识的打开摸鱼网站，插件帮助我们减少摸鱼的时间和频率，提高我们上班和学习的效率，节省时间用于学习提升自己或者享受生活**。

![](https://github.com/OBKoro1/stop-mess-around/blob/dev/static/example/run-introduction8M.gif?raw=true)

#### [AutoCommit](https://github.com/OBKoro1/autoCommit)

这是一个用于Git自动commit的VSCode插件，它可以用来补充之前忘记提交commit，帮助你把首页的绿色格子填满。

![](https://github.com/OBKoro1/autoCommit/raw/master/images/autoCommit.gif?raw=true)

### [web-basics](https://github.com/OBKoro1/web-basics)

收集和整理了一个大厂前端需要掌握能力的仓库。

其中分为JS基础能力，大厂场景题、大厂面试真题。

希望能够帮助大家提升自己的能力，在面试的时候能够游刃有余，轻松拿到高薪offer。

![大厂前端需要掌握的能力](https://github.com/OBKoro1/web-basics/blob/main/static/web-basic-example.gif?raw=true)

### 用爱发电，求赞助 😭

开源不易，本插件的开发与维护全都是利用业余时间。

如果觉得这个效率工具还不错, 对你有所帮助，就赞助支持一下我的工作吧。

请我喝杯水吧，十块八块不嫌多，三块五块不嫌少 ❤️

<img src="https://raw.githubusercontent.com/OBKoro1/koro1FileHeader/dev/images/pay.jpg" alt="赞助" width="500px" style="display: block;"/>


### License

[MIT](http://opensource.org/licenses/MIT)

### 求Star

如果觉得还不错的话，就给个 [Star](https://github.com/OBKoro1/koro1FileHeader) ⭐️ 鼓励一下我吧~

## 联系我

[掘金](https://juejin.im/user/78820536236951)、[前端进阶积累](http://obkoro1.com/web_accumulate/)、[公众号](https://user-gold-cdn.xitu.io/2018/5/1/1631b6f52f7e7015?w=344&h=344&f=jpeg&s=8317)、[GitHub](https://github.com/OBKoro1)、[微信](https://raw.githubusercontent.com/OBKoro1/articleImg_src/master/weibo_img_move/005Y4rCogy1fsnslyz5pnj309j0cdgm6.jpg):OBkoro1、邮箱：obkoro1@foxmail.com



CHANGELOG.md
<!--
 * Author       : OBKoro1
 * Date         : 2021-03-27 17:30:30
 * Last Author  : OBKoro1 1677593011@qq.com
 * LastEditTime : 2023-01-29 16:04:10
 * FilePath     : /koro1FileHeader/CHANGELOG.md
 * Description  :
-->

# 更新日志

### 如果觉得本插件还不错的话，给个[Star](https://github.com/OBKoro1/koro1FileHeader)吧~

## 文档

[配置文档](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE%E5%AD%97%E6%AE%B5) - 有所有功能的简介的 json 文件

[配置](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE) - 所有配置的详细描述的文档，用于查看详细介绍。

### [V4.9.3]

* feat: specialOptions支持单独根据语言或者语言后缀进行配置，参考自定义语言的后缀形式 [#539](https://github.com/OBKoro1/koro1FileHeader/issues/539)， 文档：[特殊字段允许自定义](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE#6-%E7%89%B9%E6%AE%8A%E5%AD%97%E6%AE%B5%E5%85%81%E8%AE%B8%E8%87%AA%E5%AE%9A%E4%B9%89)

如下面的js、md，会针对后缀文件进行单独的特殊字段设置。

```js
"fileheader.configObj": {
  "specialOptions":{
    "Author": "creater",
    "Date": "since",
    "LastEditTime": "lastTime",
    "LastEditors": "LastAuthor",
    "Description": "message", // 头部注释大写的描述Description
    "description": "function message", // 函数注释小写的描述：description
    "FilePath": "文件相对于项目的路径"
    "param": "param2", // 函数注释parm参数别名
    // 文件后缀、或者语言后缀，可针对单个文件后缀进行配置：language的自定义语言配置
    "js": {
        "Description": "message2",  // 合并specialOptions第一层级的配置，覆盖Description的配置
        "return": "return2", // 单独为js文件添加return 特殊字段配置
    },
    "md": {
      "Date": "date",
      "Description": "tag"
    }
  }
}
```


* feat: 函数注释特殊字段为description,头部注释的特殊字段为Description，用于区分 [#512](https://github.com/OBKoro1/koro1FileHeader/issues/512)

文档：[特殊字段允许自定义](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE#6-%E7%89%B9%E6%AE%8A%E5%AD%97%E6%AE%B5%E5%85%81%E8%AE%B8%E8%87%AA%E5%AE%9A%E4%B9%89)、[移动光标到Description :所在行](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE#%E7%A7%BB%E5%8A%A8%E5%85%89%E6%A0%87%E5%88%B0description-%E6%89%80%E5%9C%A8%E8%A1%8C)



* fix: now_year改为全局匹配 [#522](https://github.com/OBKoro1/koro1FileHeader/issues/522)
* fix: now_year、用户名和邮箱不再绑定 [#509](https://github.com/OBKoro1/koro1FileHeader/issues/509)
* fix: param等宽错误  [#495](https://github.com/OBKoro1/koro1FileHeader/issues/495)
* fix: 修复vscode默认快捷键和插件简介描述 [#498](https://github.com/OBKoro1/koro1FileHeader/issues/498)

### [V4.9.2]

* fix: VSCode 1.74.0 兼容，解决函数注释无法使用的问题

### [V4.9.1]

* fix: 自定义特殊字段LastEditors取值导致的问题 [#478](https://github.com/OBKoro1/koro1FileHeader/issues/477)
* fix: 修复params读取失败的问题 [#477](https://github.com/OBKoro1/koro1FileHeader/issues/477)
* fix: 修复git config读取失败的问题 [#475](https://github.com/OBKoro1/koro1FileHeader/issues/475)

### [V4.9.0]

- feat: `NoMatchParams`当没有匹配到函数注释参数时是否显示`param`和`return`

```js
"fileheader.configObj": {
  "NoMatchParams": "no param" // 默认不显示param那一行
}
```

示例:

```js
// "NoMatchParams": "no show param"
/**
 * @description: 没匹配到函数参数，不显示@param那行
 * @return {type} 填写返回值
 */

export const download = async () => {
  // do something
  console.log('空参数 匹配不到函数参数: NoMatchParams')
  return 'something'
}

// "NoMatchParams": "no show param and return"
/**
 * @description: 只显示description 不显示@param与@return这两行
 */
export const download = async () => {
  // do something
  console.log('空参数 匹配不到函数参数: NoMatchParams')
  return 'something'
}

// "NoMatchParams": "show param"
/**
 * @description: 没匹配到函数参数 也显示@param与@return
 * @param {type}
 * @return {type}
 */
export const download = async () => {
  console.log('空参数 匹配不到函数参数: NoMatchParams')
}
```

### [V4.8.23]

- feat: `functionParamAddStr`在`type param`后面增加字符串, 在 type param 后面增加字符串 可能是冒号，方便输入参数描述。 [#443](https://github.com/OBKoro1/koro1FileHeader/issues/443)

```js
"fileheader.configObj": {
  "functionParamAddStr": "" // 默认不增加字符串
}
```

示例:

```js
// "functionParamAddStr": ":"
/**
 * @description: 增加冒号 方便输入参数描述
 * @param {type} a:
 * @param {type} b:
 * @param {array} c:
 * @return {type}
 */
function test2(a, b, ...c) {}
```

- fix: 在 VSCode 打开的对应项目中获取`git config`信息 [#469](https://github.com/OBKoro1/koro1FileHeader/issues/469)
- fix: 修复`java`函数注释参数提取不精准的问题 [#470](https://github.com/OBKoro1/koro1FileHeader/issues/470)
- fix: 修复不设置`Author`和`LastEditors`时出现的无法生成注释的 bug
- fix: c、java、solidity 不支持连续的容器对象作为参数 [#468](https://github.com/OBKoro1/koro1FileHeader/issues/468)

### [V4.8.22]

- feat: [新增头部注释模板`Author`和`LastEditors`从 git config 中读取`user.name`、`user.email`的功能](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE#author%E5%92%8Clasteditors%E8%8E%B7%E5%8F%96git-config%E9%85%8D%E7%BD%AE)。
- feat: [新增版权声明获取 git config 配置: `${git_name}、${git_email}、${git_name_email}`](https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE#%E7%89%88%E6%9D%83%E5%A3%B0%E6%98%8E%E4%B9%9F%E6%94%AF%E6%8C%81%E8%AF%BB%E5%8F%96git-config%E9%85%8D%E7%BD%AE)

配置:

```js
// 头部注释模板
"fileheader.customMade": {
  // 获取当前项目下git config中的: 用户名、邮箱
  // 公司项目和个人项目可以配置不同的用户名与邮箱 搜索: gitconfig includeIf  比如: https://ayase.moe/2021/03/09/customized-git-config/
  "Author": "git config user.name && git config user.email", // 同时获取用户名与邮箱
  // "Author": "git config user.name", // 仅获取用户名
  // "Author": "git config user.email", // 仅获取邮箱
  // "Author": "OBKoro1", // 固定值 不从git config中获取
  "Date": "Do not edit",
  // 设置后，保存文件更改默认更新最后编辑人
  "LastEditors": "git config user.name && git config user.email", // 配置规则与Author字段一致
  "LastEditTime": "Do not edit",
  "FilePath": "",
  // 个人版权声明: ${git_name} ${git_email} ${git_name_email}
  // 公司版权声明: 写死即可
  "custom_string_obkoro1_copyright": "Copyright (c) ${now_year} by ${git_name} email: ${git_email}, All Rights Reserved. "
},
```

示例:

```js
// "Author": "git config user.name && git config user.email"
// "LastEditors": "git config user.name"
//  "custom_string_obkoro1_copyright": "Copyright (c) ${now_year} by ${git_name} email: ${git_email}, All Rights Reserved. "
/*
 * Author       : OBKoro1 obkoro1@foxmail.com
 * Date         : 2022-05-01 15:35:12
 * LastEditors  : OBKoro1
 * LastEditTime : 2022-05-02 11:31:09

src/logic.js
/*
 * @Description: 逻辑输出
 * @Author     : OBKoro1
 * @Date       : 2018-10-31 16:22:55
 * @LastAuthor : OBKoro1
 * @lastTime   : 2019-08-08 16:32:27
 */
const vscode = require('vscode');
const languageOutput = require('./languageOutput');
const util = require('./util');
const fs = require('fs');
const path = require('path');
const global = require('./CONST');

/**
 * @description: 头部注释根据用户设置返回模板数据对象
 * @param {Object} userObj 用户设置
 * @param {String} time 文件创建时间
 * @return: 返回生成模板的数据对象
 */
const userSet = config => {
  const userObj = config.customMade;

  let data = {};
  let userSet = Object.keys(userObj);
  if (userSet.length === 0) {
    // 默认模板
    data = {
      Author: 'your name',
      Date: '',
      LastEditTime: '',
      LastEditors: 'your name',
      Description: 'In User Settings Edit',
      FilePath: ''
    };
  } else {
    // 如果用户设置了模板，那将默认根据用户设置模板
    data = Object.assign({}, userObj); // 复制对象，否则对象不能更改值
  }
  data = changeDataOptionFn(data, config);
  return data;
};

/**
 * @description: 函数注释前面的长度
 * @param {Object} editor 当前激活文件
 * @return: lineSpace：前面的长度，frontStr：函数注释第一行的长度，line:当前行(数字)，nextLine 激活行的下一行是否有内容
 */
const lineSpaceFn = editor => {
  const line = editor.selection.active.line; // 当前行
  const lineProperty = editor.document.lineAt(line); // 当前行的属性
  let lineFirst = lineProperty.firstNonWhitespaceCharacterIndex; // 激活行 前面是否有值
  let lineSpace,
    nextLine,
    frontStr = ''; // 前面空几行
  const activeLine = editor.selection.active.line; // 激活行 行号
  // 判断当前行有没有内容 决定选择当前行还是下一行的长度
  if (
    lineProperty.isEmptyOrWhitespace &&
    editor._documentData.document.lineCount !== activeLine + 1
  ) {
    nextLine = activeLine + 1;
    lineSpace = editor.document.lineAt(nextLine)
      .firstNonWhitespaceCharacterIndex;
    lineFirst = lineFirst === 0 ? lineSpace : 0;
    frontStr = ''.padStart(lineFirst);
  } else {
    lineSpace = lineFirst;
  }
  return [lineSpace, frontStr, line, nextLine];
};

/**
 * @description: 保存时触发修改
 * @param {object} document 文档对象
 * @param {object} userObj 用户设置
 * @param {String} fileEnd 文件后缀
 * @return: authorRange 原修改人行
 * @return: authorText  当前修改人
 * @return: lastTimeRange  原最后编辑时间
 * @return: lastTimeText 当前编辑时间
 * @return: hasAnnotation 是否自动添加头部注释
 */
function saveReplaceTime(document, config, fileEnd) {
  const userObj = config.customMade;
  let authorRange, authorText, lastTimeRange, lastTimeText;
  let changeFont = new languageOutput.changeFont(fileEnd);
  let annotationStarts = changeFont.star();
  let totalLine = document.lineCount - 1; // 总行数
  let enter = false;
  let hasAnnotation = false; // 默认没有
  // 有没有更改特殊变量
  const checkHasAnnotation = (name, line, checked) => {
    if (checked) return false; // 已经找到要替换的
    let userSetName = config.configObj.specialOptions[name];
    if (userSetName) {
      if (line.indexOf(`${userSetName}`) === -1) {
        // 没有检测用户自己更改的 再检测特殊变量
        return line.indexOf(`${name}`) !== -1;
      } else {
        // 检测用户自己更改的
        return true;
      }
    } else {
      // 检测特殊变量
      return line.indexOf(`${name}`) !== -1;
    }
  };

  for (let i = 0; i < 15; i++) {
    // 只遍历前15行没有文件头部注释内容即退出
    let linetAt = document.lineAt(i); // 获取每行内容
    let lineNoTrim = linetAt.text; // line
    let line = linetAt.text.trim();
    if (!enter) {
      // 判断进入注释
      if (annotationStarts === line || annotationStarts === lineNoTrim) {
        enter = true;
      }
    } else {
      let range = linetAt.range;
      if (checkHasAnnotation('LastEditors', line, authorRange)) {
        //表示是修改人
        hasAnnotation = true;
        authorRange = range;
        let LastEditors = userObj.LastEditors || 'Please set LastEditors';
        authorText = changeFont.LastEditorsStr(LastEditors);
      } else if (checkHasAnnotation('LastEditTime', line, lastTimeRange)) {
        //最后修改时间
        hasAnnotation = true;
        lastTimeRange = range;
        lastTimeText = changeFont.lastTimeStr();
      } else if (checkHasAnnotation('Date', line)) {
        hasAnnotation = true;
      }
    }
    if (totalLine === i) break; // 行数不够则退出循环
  }
  return [authorRange, authorText, lastTimeRange, lastTimeText, hasAnnotation];
}

/**
 * @description: 取出文件后缀
 * @param { String } 文件后缀
 * @param {Object} config 用户设置
 * @return: [生成注释的行数,注释之前添加的内容,注释之前添加的内容]
 */
const editLineFn = (fsPath, config) => {
  const pathArr = fsPath.split('/');
  const fileName = pathArr[pathArr.length - 1];
  const fileNameArr = fileName.split('.');
  let fileEnd = fileNameArr[fileNameArr.length - 1]; // 文件后缀
  let isSpecial = util.specialLanguageFn(fsPath, config);
  // 特殊文件
  if (isSpecial) {
    fileEnd = isSpecial;
  }

  // 切割文件路径 获取文件后缀
  const headInsertLineObj = config.configObj.headInsertLine;
  let lineNum = 0;
  if (headInsertLineObj[fileEnd]) {
    lineNum = headInsertLineObj[fileEnd] - 1;
  }
  // 是否设置在注释之前添加内容
  let isSetAdd = config.configObj.beforeAnnotation[fileEnd];
  let isAfterAdd = config.configObj.afterAnnotation[fileEnd];
  return [lineNum, isSetAdd, isAfterAdd];
};

// 将字段弄得一样长
const sameLengthFn = (data, maxNum) => {
  let objData = {};
  // 修改属性
  Object.keys(data).forEach(item => {
    let diffNum = maxNum - item.length;
    // 负值和0不改变
    if (diffNum < 1) {
      objData[item] = data[item];
    } else {
      const newItem = util.spaceStringFn(item, maxNum);
      objData[newItem] = data[item];
    }
  });
  return objData;
};

/**
 * 更改字段，不改变他们的顺序
 * @Created_time: 2019-05-07 19:36:20
 * @return {Object} 更换字段后的对象
 */
const changePrototypeNameFn = (data, config) => {
  let keysArr = Object.keys(data);
  let specialOptions = config.configObj.specialOptions; // 时间字段重命名配置
  let objData = {};
  let specialArr = [
    'Date',
    'LastEditTime',
    'LastEditors',
    'Description',
    'FilePath'
  ];
  keysArr.forEach(item => {
    // 特殊字段 且 有设置特殊字段
    if (specialArr.includes(item) && specialOptions[item]) {
      objData[specialOptions[item]] = data[item];
    } else if (item === 'custom_string_obkoro1') {
      // 更改用户自定义输出字段 后期需要切割它
      objData.symbol_custom_string_obkoro1 = data[item];
    } else {
      objData[item] = data[item];
    }
  });
  return objData;
};
/**
 * 修改时间，描述等配置值
 * @param {object} data 配置项
 */
function changeDataOptionFn(data, config) {
  let time = new Date().format();
  // 文件创建时间
  if (config.configObj.createFileTime) {
    const filePath =
      vscode.window.activeTextEditor._documentData._document.fileName;
    time = new Date(fs.statSync(filePath).birthtime).format();
  }
  // 判断是否设置
  if (data.Date !== undefined) {
    data.Date = time;
  }
  // 当前时间为最后编辑时间
  if (data.LastEditTime !== undefined) {
    data.LastEditTime = new Date().format();
  }
  // 自动添加文件路径
  if (data.FilePath !== undefined) {
    let { itemName, fileItemPath } = util.getFileRelativeSite();
    let res = `${path.sep}${itemName}${fileItemPath}`; // 拼接项目名称和相对于项目的路径
    if (data.FilePath === 'no item name') {
      res = `${fileItemPath}`;
    }
    if (config.configObj.filePathColon !== '路径分隔符替换') {
      // path.sep window: \ mac: /
      const reg = new RegExp(path.sep, 'y');
      res = res.replace(reg, config.configObj.filePathColon);
    }
    data.FilePath = res;
  }
  data = changePrototypeNameFn(data, config);
  if (config.configObj.wideSame) {
    data = sameLengthFn(data, config.configObj.wideNum);
  }
  return data;
}

/**
 * 函数注释，更改值,
 * @Created_time: 2019-05-07 19:36:20
 * @return {Object} 更换字段后的对象
 */
const cursorOptionHandleFn = config => {
  let data = {};
  let userSet = Object.keys(config.cursorMode);
  if (userSet.length === 0) {
    data = {
      description: '',
      param: '',
      return: ''
    };
  } else {
    // 如果用户设置了模板，那将默认根据用户设置模板
    data = Object.assign({}, config.cursorMode); // 复制对象，否则对象不能更改值
  }
  if (data.Date !== undefined) {
    data.Date = new Date().format();
  }
  data = changNameFn(data, config);
  return data;
};

/**
 * 更改字段，不改变他们的顺序
 * @param {obeject} data 函数模板配置
 * @param {*} config 顶层配置
 */
function changNameFn(data, config) {
  let keysArr = Object.keys(data);
  let specialOptions = config.configObj.specialOptions; // 时间字段重命名配置
  let objData = {};
  // 支持日期和描述
  let specialArr = ['Date', 'Description'];
  keysArr.forEach(item => {
    if (specialArr.includes(item) && specialOptions[item]) {
      // 特殊字段重新赋值
      objData[specialOptions[item]] = data[item];
    } else if (item === 'custom_string_obkoro1') {
      objData.symbol_custom_string_obkoro1 = data[item];
    } else {
      objData[item] = data[item];
    }
  });
  return objData;
}

/**
 * @description: 处理生成的模板 比如添加信息，删除信息等。
 * @param {Object} beforehand 模板和预处理的参数
 * beforehand.tpl {string} tpl 模板changePrototypeNameFn
 * beforehand.beforeAnnotation {Boolean}  是否在模板之前添加内容
 * beforehand.afterAnnotation {Boolean}  是否在模板之后添加内容
 * beforehand.fileEnd 文件后缀
 * @return: {String} tpl
 * @Created_time: 2019-05-14 14:25:26
 */
const handleTplFn = beforehand => {
  let res = util.replaceSymbolStr(beforehand.tpl, beforehand.fileEnd);
  if (beforehand.beforeAnnotation) {
    res = `${beforehand.beforeAnnotation}\n${res}`;
  }
  if (beforehand.afterAnnotation) {
    res = `${res}${beforehand.afterAnnotation}\n`;
  }
  return res;
};

// 自动添加是否匹配黑名单
const isMatchProhibit = (fsPath, config) => {
  let match = false;
  let prohibit = config.configObj.prohibitAutoAdd;
  let fsName = util.fsPathFn(fsPath);
  if (prohibit && prohibit.length > 0) {
    match = prohibit.includes(fsName);
  }
  return match;
};

/**
 * @description: 逻辑判断
 * @param {Object} params
 * @param {Number} params.lineCount 文件行数
 * @param {String} params.fsPath 文件路径
 * @param {String} params.fileEnd 文件后缀
 * @param {Boolean} params.hasAnnotation 文件是否已有头部注释
 * @param {Object} params.config 插件配置
 * @Created_time: 2019-11-02 17:12:51
 * @return {Boolean} 是否自动添加
 */
const isAutoAddFn = params => {
  // 文件超过一定行数时 不自动添加头部注释
  if (params.config.configObj.autoAddLine < params.lineCount) return false;
  if (!params.config.configObj.autoAdd) return false; // 关闭自动添加
  if (params.hasAnnotation) return false; // 文件已经有注释
  const hasAddProhibit = isMatchProhibit(params.fsPath, params.config);
  if (hasAddProhibit) return false; // 被添加进黑名单
  // 曾经自动添加过头部注释 不再添加
  if (global.autoAddFiles.includes(params.fsPath)) return false;
  if (
    params.config.configObj.autoAlready &&
    params.fileEnd === '匹配不到_默认注释'
  ) {
    // 只自动添加支持的语言 该文件不是插件支持的语言
    return false;
  }
  return true; // 自动添加
};

/**
 * @description: 函数注释移动光标到description所在行
 * @param {String} tpl 最终要生成的模板
 * @Created_time: 2019-06-18 14:28:13
 */
const moveCursorDesFn = (fileEnd, config, fontTpl, lineNum) => {
  // 生成Description行
  if (!config.configObj.moveCursor) return;
  const editor = vscode.editor || vscode.window.activeTextEditor; // 每次运行选中文件
  const specialOptions = config.configObj.specialOptions; // 时间字段重命名配置
  const DescriptionName = specialOptions.Description
    ? specialOptions.Description
    : 'Description';
  let data = {
    [DescriptionName]: ''
  };
  let str = languageOutput.middleTpl(data, fileEnd, config);
  str = str.trim();
  // 计算函数注释模板行数
  let newLineNum = fontTpl.split(/\r\n|\r|\n/).length - 1;
  let i = lineNum - 1; // 初始行数
  let descriptionLineNum; // 目标行
  for (i < i + newLineNum; i++; ) {
    let line = editor.document.lineAt(i);
    let lineNoTrim = line.text; // line
    if (lineNoTrim.indexOf(str) !== -1) {
      descriptionLineNum = i;
      break;
    }
    if (editor.document.lineCount - 1 === i) break; // 总行数
  }
  // 没有Description 则不移动视图
  if (descriptionLineNum === undefined) {
    return;
  }
  // 移动光标到指定行数
  const position = editor.selection.active;
  var newPosition = position.with(descriptionLineNum, 10000);
  editor.selection = new vscode.Selection(newPosition, newPosition);
};

/**
 * @description: 移动光标到description所在行 移动视图到顶部
 * @param {String} tpl 最终要生成的模板
 * @Created_time: 2019-06-18 14:28:13
 */
const moveCursor = tpl => {
  const config = vscode.workspace.getConfiguration('fileheader'); // 配置项默认值
  if (config.configObj.moveCursor) {
    const editor = vscode.editor || vscode.window.activeTextEditor; // 每次运行选中文件
    const specialOptions = config.configObj.specialOptions; // 时间字段重命名配置
    const DescriptionName = specialOptions.Description
      ? specialOptions.Description
      : 'Description';
    // 注释总行数 最后多一行注释开头 一行注释结尾 最后一行换行
    const strLine = tpl.split(/\r\n|\r|\n/).length;
    // 文档是从0开始 行数从1开始 要减去1
    let descriptionLineNum;
    for (let i = 0; i < strLine; i++) {
      let line = editor.document.lineAt(i);
      let lineNoTrim = line.text; // line
      if (lineNoTrim.indexOf(`${DescriptionName}:`) !== -1) {
        descriptionLineNum = i;
        break;
      }
      if (editor.document.lineCount - 1 === i) break; // 总行数
    }
    // 没有Description 则不移动视图
    if (descriptionLineNum === undefined) {
      return;
    }
    // 移动光标到指定行数
    const position = editor.selection.active;
    var newPosition = position.with(descriptionLineNum, 10000);
    editor.selection = new vscode.Selection(newPosition, newPosition);
    // 移动视图到顶部
    vscode.commands.executeCommand('editorScroll', {
      to: 'up',
      value: 10000
    });
  }
};

module.exports = {
  userSet,
  lineSpaceFn,
  saveReplaceTime,
  cursorOptionHandleFn,
  editLineFn,
  handleTplFn,
  isAutoAddFn,
  moveCursor,
  moveCursorDesFn
};
